File: Backtracking/M-40. Combination Sum II/M-40. Combination Sum II.py
from typing import List


class Solution:
    @classmethod
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        result: list[int] = []
        candidates.sort()

        def backtracking(cur_idx: int, rem_target: int, cur_candidates: list[int]):
            if rem_target == 0:
                result.append(cur_candidates.copy())
                return

            if rem_target < 0:
                return

            for i in range(cur_idx, len(candidates)):
                candidate = candidates[i]

                if i > 0 and candidates[i - 1] == candidate:
                    continue

                cur_candidates.append(candidate)
                candidates.remove(candidate)
                backtracking(cur_idx=i,
                             rem_target=rem_target - candidate,
                             cur_candidates=cur_candidates
                             )
                candidates.insert(i, candidate)
                cur_candidates.pop()

        backtracking(cur_idx=0,
                     rem_target=target,
                     cur_candidates=[]
                     )

        return result


test_data = [
    {
        "candidates":  [10, 1, 2, 7, 6, 1, 5],
        "target": 8,
        "ans": [
            [1, 1, 6],
            [1, 2, 5],
            [1, 7],
            [2, 6],
        ],
    },
    {
        "candidates": [2, 5, 2, 1, 2],
        "target": 5,
        "ans": [[1, 2, 2], [5]],
    },
    {
        "candidates": [2],
        "target": 1,
        "ans": [],
    },
]

for data in test_data:
    print(
        data["ans"] == Solution.combinationSum2(
            data["candidates"], data["target"])
    )

File: Backtracking/M-40. Combination Sum II/M-40. Combination Sum II.ts
function combinationSum2(candidates: number[], target: number): number[][] {
  const result: number[][] = [];
  candidates.sort((a, b) => a - b);
  const backtracking = (
    curIdx: number,
    remTarget: number,
    curCandidates: number[]
  ): void => {
    if (remTarget === 0) {
      result.push([...curCandidates]);
      return;
    }

    if (remTarget < 0) return;

    for (let i = curIdx; i < candidates.length; i++) {
      const candidate = candidates[i];
      /** to prevent we starting from a same root */
      if (i > 0 && candidate === candidates[i - 1]) continue;

      curCandidates.push(candidate);
      candidates.splice(i, 1);
      backtracking(i, remTarget - candidate, curCandidates);
      candidates.splice(i, 0, candidate);
      curCandidates.pop();
    }
  };

  backtracking(0, target, []);
  return result;
}

const testData = [
  {
    candidates: [10, 1, 2, 7, 6, 1, 5],
    target: 8,
    ans: [
      [1, 1, 6],
      [1, 2, 5],
      [1, 7],
      [2, 6],
    ],
  },
  {
    candidates: [2, 5, 2, 1, 2],
    target: 5,
    ans: [[1, 2, 2], [5]],
  },
  {
    candidates: [2],
    target: 1,
    ans: [],
  },
];

for (const data of testData) {
  console.log(
    JSON.stringify(data.ans) ===
      JSON.stringify(combinationSum2(data.candidates, data.target))
  );
}
